Add functionality to generate TypeScript definitions

JamesSkemp opened the pull request:
Based upon local testing (`npm install file:../..`), these changes appear to add proper TypeScript definitions to the library.

This change consists of the following:

1. Add TypeScript as a devDependency.
2. Add a new script to leverage TypeScript's built-in functionality to generate .d.ts files.
    - Note that I did try to have it generate one file, but couldn't get it to work, and looking at other projects they either are built with TypeScript, or are simpler. My hope is that adding some TS support will lead to someone with more knowledge than I improving the generation process. Or the library being rewritten in TypeScript. ;)
3. The actual definitions files, after running the new `npm run declaration:build` process.
    - This should probably be added to the build process so that it's updated whenever the package is updated/published.

Based upon local testing, this then allows you to add the rpg-dice-roller package and import functionality (`import { DiceRoll } from 'rpg-dice-roller';`).

This relates to and would close #92 .

GreenImp commented:
Hey, this looks really good!
I've got a few questions, having not touched Typescript before, but also some bad news.. All PRs need to be branched off `develop` and merged into either `develop` or a feature branch, so I'm going to have to reject this one. If you're able to branch from `develop` instead of `master` and get another PR in, I'd love to see if we can get this merged.

I've left some comments on the changes themselves (In the "Files changed" tab), just some minor questions.
I'm also wondering if it would be best to not include the generated definition files in the git repo, as they can be built as part of the publishing process.

JamesSkemp commented:
Durr. I should have remembered that from my previous PR. I wish GitHub would keep the original's default branch when you fork ...

I've updated my fork with develop changes, and will get my changes cherry picked or manually back in to a new PR.

And no, the types directory could easily be added to the .gitignore, since you're absolutely correct that they're built. I'm not seeing any inline comments on the files, nor do I have any emails with them, but I'll see if I lost that because of the branch change. (I saw the word `develop` and immediately recalled what was wrong, and corrected that bit.)

JamesSkemp commented:
I don't suppose your comments are still pending, are they? I'm having a hard time finding them, and have double-checked that I received no emails with the comments, nor notification in GitHub (other than your comment on my question issue).

https://stackoverflow.com/a/45336490/11912

JamesSkemp commented on tsconfig.json:
```diff
@@ -0,0 +1,26 @@
+{
```
Good question.

Yes, the tsconfig.json is a normal part of a TypeScript project. If a project supports TypeScript, it will very likely have such a file. The additional file, declaration.tsconfig.json, extends upon it.

GreenImp replied:
Makes sense.

JamesSkemp commented on readme.md:
````diff
@@ -18,6 +18,10 @@ A JS based dice roller that can roll various types of dice and modifiers, along
 npm install rpg-dice-roller
 ```
 
+## Update TypeScript Definitions
````
This is removed in the new PR I'll be submitting.

GreenImp replied:
Fab cheers!

JamesSkemp commented on package.json:
```diff
@@ -68,7 +70,8 @@
     "test:coveralls": "jest /tests/ --coverage && coveralls < coverage/lcov.info",
     "prepublishOnly": "npm run build:prod && npm run build:dev && npm run test",
```
Perfect. Updated in the new PR I'll be submitting.

Would you like the `prepublishOnly` script to be the last one? If so, I can reorder the name of the script so that it's `build:declaration`, and is properly alphabetized with the existing `build:` scripts.

GreenImp replied:
Hmm that's a good thought. Might be worth renaming it to `build:declaration` or `build:definitions`, for the sake of consistency.
At the moment, they're not alphabetised, so I'm not overly concerned by that, but it might be good to put it either before or after the `build:grammars` script.

GreenImp commented:
Ah, good spot! More used to Bitbucket's reviews. I assumed the "pending" meant that it was waiting for a response / resolution from you! I've submitted the review now, so hopefully you can see the comments.

The `master` branch was set as the default, which has caused confusion with a few PRs, so I've just changed it to `develop` so hopefully it'll make things easier.
It might be good to branch off `develop`, cherry pick your changes, and raise a new PR, as you mentioned. Otherwise it makes the branch history a bit skewy, with `master` being fed into `develop`.

GreenImp commented:
Another thought; the NPM package is going to need the definition files and the `tsconfig.json` file, so people can actually use them. So you'll need to add exclusions for them to the `.npmignore` file.

I think something like this would do the job:

```
!/types/*.d.ts
!tsconfig.json
```

JamesSkemp commented:
Replaced by #152
